Add transform to add dedicated input and output

mtdowling opened the pull request:
Adds a ModelTransformer transform that ensures every shape defines a
dedicated input and output shape marked with the @input and @output
traits and uses a consistent naming strategy.

The goal of this transform is to consolidate much of the logic that has
been repeated in various AWS SDK code generators to generate synthetic
input and output shapes for operations. It provides a consisten name for
these shapes so that generators know that they can use the name as-is if
they'd like.

Because shapes are sometimes renamed with this transform, it's important
to know what the original shape ID was. This might matter for protocols
that need to know the name of the input shape when serializing a request
or response (like AWS rest-xml). To accomplish this, I added the concept
of ephemeral traits to the model that allow traits to be added to shapes
that are only defined in code rather than through a trait definition.
Ephemeral traits are not persisted when a model is serialized, but
otherwise can be used just like any other trait. When a shape is
renamed, an ephemeral trait named OriginalShapeIdTrait is added to the
shape to provide the original shape ID of the shape so that things like
code generators can query it.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

JordonPhillips commented on smithy-model/src/main/java/software/amazon/smithy/model/shapes/ModelSerializer.java:
```diff
@@ -59,7 +59,8 @@ private ModelSerializer(Builder builder) {
         }
-        traitFilter = builder.traitFilter;
+        // Never serialize ephemeral traits.
+        traitFilter = builder.traitFilter.and(FunctionalUtils.not(Trait::isEphemeral));
```
Wouldn't it be kind of dangerous to serialize a model that has these traits at all? It would no longer be usable. maybe we should throw an error rather than filtering the traits.

mtdowling replied:
No, the model is definitely usable without ephemeral traits. The point of these traits is that they’re only useful for intermediate transforms. For example, you might want to punch traits on to shapes only for a specific code generator, but the traits don’t need to be persisted. Every projection in Smithy is serialized, so I don’t want ephemeral traits to poison models.

JordonPhillips commented on smithy-model/src/main/java/software/amazon/smithy/model/traits/ephemeral/OriginalShapeIdTrait.java:
```diff
@@ -0,0 +1,59 @@
+ * <p>This is an ephemeral trait that is not defined in the semantic model,
+ * nor is it persisted when the model is serialized.
+ */
+public final class OriginalShapeIdTrait extends AbstractTrait {
```
nit: this could be a `StringTrait`

mtdowling replied:
I didn’t use a StringTrait because I didn’t think the getValue method should return a string here as this trait is only meant for defining the previous shape ID. Fair?

JordonPhillips commented on smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java:
```diff
@@ -0,0 +1,220 @@
+        if (model.getShapeIds().contains(newId)) {
+            ShapeId deconflicted = resolveConflict(newId, suffix);
+            if (model.getShapeIds().contains(deconflicted)) {
+                throw new ModelTransformException(String.format(
```
Have you run this on the AWS models to ensure it handles those cases?

mtdowling replied:
Will do

jdisanti replied:
Kinesis Analytics is a good one to test against since it has `DeleteApplication` and `DeleteApplicationOutput` operations.

Currently, smithy-rs places each generated operation input/output into its own Rust namespace so that name conflicts aren't an issue. This is illustrated in the docs for the DeleteApplication/DeleteApplicationOutput outputs in aws-sdk-kinesisanalytics:
https://docs.rs/aws-sdk-kinesisanalytics/0.0.26-alpha/aws_sdk_kinesisanalytics/output/index.html
https://docs.rs/aws-sdk-kinesisanalytics/0.0.26-alpha/aws_sdk_kinesisanalytics/operation/index.html

So if I understand these changes correctly, these would become:
- DeleteApplicationOperationOutput
- DeleteApplicationOutputOutput
Rather than:
- DeleteApplicationOutput
- DeleteApplicationOutputOutput

But the change in name is kind of moot for smithy-rs since each of these get their own Rust namespace anyway. I think I prefer leaving them as is and placing the synthetic shapes in a synthetic Smithy namespace. What do you think about making this configurable to that each code generator can decide how the conflict should be resolved?

mtdowling replied:
If you use Request/Response as your suffix, it becomes `DeleteApplicationOutputRequest` / `DeleteApplicationOutputResponse`.

And yeah, you're correct; if you use Input/Output, it becomes:
* `DeleteApplication`: `DeleteApplicationOperationOutput`
* `DeleteApplicationOutput`: `DeleteApplicationOutputOutput`

> But the change in name is kind of moot for smithy-rs since each of these get their own Rust namespace anyway. I think I prefer leaving them as is and placing the synthetic shapes in a synthetic Smithy namespace.

If you want to move input and output shapes to another namespace, then you can do that by first performing the transform, then iterating over every structure shape marked with the `@input` and `@output` shapes, and generate those into whatever namespace/package you want. You can use the applied `OriginalShapeIdTrait` if you want to further change the name (this could be added to your symbol provider for example). The transform is still useful because it will copy shapes when they're reused within inputs and outputs and it will remove unreferenced input and output shapes after copies are made.

> What do you think about making this configurable so that each code generator can decide how the conflict should be resolved?

We can't give you the exact names you have in Rust using a transform like this. Every shape, including operations, need to have unique names in the closure of a service, so you can't reuse `DeleteApplicationOutput` as both an operation name and output shape name.

jdisanti replied:
I see. Makes sense.

JordonPhillips commented on smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    private static ShapeId resolveConflict(ShapeId id, String suffix) {
+        // Say GetFooRequest exists. This then returns GetFooOperationRequest.
```
Why change deconflict onto the input / output name rather than renaming the shape that conflicts? I ask this because doing it this way means that the input / output shapes don't have consistent names.

mtdowling replied:
The shape name that conflicts is arguably more important to have a consistent name. For example, they might appear in things like document types in the future, whereas the input/output traits would not allow the input/output shapes to appear there. But also… you could say the same thing about renaming inner shapes instead of input/output shapes— why rename those, and what strategy would you use to deconflict? These renames will be rare too.

JordonPhillips commented on smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java:
```diff
@@ -0,0 +1,220 @@
+        return totalDirectedEdges == 1;
+    }
+
+    private static StructureShape createSyntheticShape(
```
I don't think this would work with the `Unit` type since the unitType trait can only be applied to it.

mtdowling replied:
Great catch. Fixed.

JordonPhillips commented on smithy-model/src/test/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutputTest.java:
```diff
@@ -0,0 +1,95 @@
+        Node actualNode = ModelSerializer.builder().build().serialize(actualResult);
+        Node expectedNode = ModelSerializer.builder().build().serialize(expectedResult);
+
+        Node.assertEquals(actualNode, expectedNode);
```
Since the original id trait isn't serialized, you're not asserting in these tests that it's applied at all

mtdowling replied:
Very good catch. I actually had a couple misses because of this.

JordonPhillips reviewed (commented):
Looks good, contingent upon the test run across current AWS models. I don't *think* there would be any issues, but a double check would be nice.

mtdowling commented:
The transform on AWS models looks good to me. I made some changes to Smithy-Diff to clean the report up too, so that was a good suggestion. Here's the report for API Gateway V2:

<details>
  <summary>API Gateway diff report</summary>
<pre>
[NOTE] com.amazonaws.apigatewayv2#CreateIntegrationOperationResponse: Added structure `com.amazonaws.apigatewayv2#CreateIntegrationOperationResponse` | AddedShape apigatewayv2.2018-11-29.json:1829:59
[NOTE] com.amazonaws.apigatewayv2#DeleteRouteResponseResponse: Added structure `com.amazonaws.apigatewayv2#DeleteRouteResponseResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#UpdateRouteOperationResponse: Added structure `com.amazonaws.apigatewayv2#UpdateRouteOperationResponse` | AddedShape apigatewayv2.2018-11-29.json:9376:53
[NOTE] com.amazonaws.apigatewayv2#DeleteModelResponse: Added structure `com.amazonaws.apigatewayv2#DeleteModelResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteRouteOperationResponse: Added structure `com.amazonaws.apigatewayv2#DeleteRouteOperationResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteDomainNameResponse: Added structure `com.amazonaws.apigatewayv2#DeleteDomainNameResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteDeploymentResponse: Added structure `com.amazonaws.apigatewayv2#DeleteDeploymentResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#CreateRouteOperationResponse: Added structure `com.amazonaws.apigatewayv2#CreateRouteOperationResponse` | AddedShape apigatewayv2.2018-11-29.json:2340:53
[NOTE] com.amazonaws.apigatewayv2#DeleteAccessLogSettingsResponse: Added structure `com.amazonaws.apigatewayv2#DeleteAccessLogSettingsResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteApiMappingResponse: Added structure `com.amazonaws.apigatewayv2#DeleteApiMappingResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#UntagResourceResponse: Added structure `com.amazonaws.apigatewayv2#UntagResourceResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteCorsConfigurationResponse: Added structure `com.amazonaws.apigatewayv2#DeleteCorsConfigurationResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteIntegrationResponseResponse: Added structure `com.amazonaws.apigatewayv2#DeleteIntegrationResponseResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteAuthorizerResponse: Added structure `com.amazonaws.apigatewayv2#DeleteAuthorizerResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#ResetAuthorizersCacheResponse: Added structure `com.amazonaws.apigatewayv2#ResetAuthorizersCacheResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteIntegrationOperationResponse: Added structure `com.amazonaws.apigatewayv2#DeleteIntegrationOperationResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteRouteRequestParameterResponse: Added structure `com.amazonaws.apigatewayv2#DeleteRouteRequestParameterResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#DeleteRouteSettingsResponse: Added structure `com.amazonaws.apigatewayv2#DeleteRouteSettingsResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#GetRouteOperationResponse: Added structure `com.amazonaws.apigatewayv2#GetRouteOperationResponse` | AddedShape apigatewayv2.2018-11-29.json:5599:50
[NOTE] com.amazonaws.apigatewayv2#DeleteStageResponse: Added structure `com.amazonaws.apigatewayv2#DeleteStageResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#UpdateIntegrationOperationResponse: Added structure `com.amazonaws.apigatewayv2#UpdateIntegrationOperationResponse` | AddedShape apigatewayv2.2018-11-29.json:8845:59
[NOTE] com.amazonaws.apigatewayv2#DeleteApiResponse: Added structure `com.amazonaws.apigatewayv2#DeleteApiResponse` | AddedShape jar:file:/Users/dowling/projects/smithy/smithy-model/build/libs/smithy-model-1.14.1.jar!/software/amazon/smithy/model/loader/prelude.smithy:57:1
[NOTE] com.amazonaws.apigatewayv2#GetIntegrationOperationResponse: Added structure `com.amazonaws.apigatewayv2#GetIntegrationOperationResponse` | AddedShape apigatewayv2.2018-11-29.json:4940:56
[ERROR] com.amazonaws.apigatewayv2#UpdateRoute: Changed operation output of `com.amazonaws.apigatewayv2#UpdateRoute` from `com.amazonaws.apigatewayv2#UpdateRouteResult` to `com.amazonaws.apigatewayv2#UpdateRouteOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:9113:47
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteRequestParameter: Changed operation output of `com.amazonaws.apigatewayv2#DeleteRouteRequestParameter` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteRouteRequestParameterResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3248:63
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteSettings: Changed operation output of `com.amazonaws.apigatewayv2#DeleteRouteSettings` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteRouteSettingsResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3350:55
[ERROR] com.amazonaws.apigatewayv2#DeleteDeployment: Changed operation output of `com.amazonaws.apigatewayv2#DeleteDeployment` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteDeploymentResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2990:52
[ERROR] com.amazonaws.apigatewayv2#DeleteIntegrationResponse: Changed operation output of `com.amazonaws.apigatewayv2#DeleteIntegrationResponse` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteIntegrationResponseResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3111:61
[ERROR] com.amazonaws.apigatewayv2#DeleteAccessLogSettings: Changed operation output of `com.amazonaws.apigatewayv2#DeleteAccessLogSettings` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteAccessLogSettingsResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2788:59
[ERROR] com.amazonaws.apigatewayv2#DeleteAuthorizer: Changed operation output of `com.amazonaws.apigatewayv2#DeleteAuthorizer` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteAuthorizerResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2912:52
[ERROR] com.amazonaws.apigatewayv2#DeleteStage: Changed operation output of `com.amazonaws.apigatewayv2#DeleteStage` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteStageResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3401:47
[ERROR] com.amazonaws.apigatewayv2#ResetAuthorizersCache: Changed operation output of `com.amazonaws.apigatewayv2#ResetAuthorizersCache` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#ResetAuthorizersCacheResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:7076:57
[ERROR] com.amazonaws.apigatewayv2#DeleteIntegration: Changed operation output of `com.amazonaws.apigatewayv2#DeleteIntegration` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteIntegrationOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3068:53
[ERROR] com.amazonaws.apigatewayv2#GetRoute: Changed operation output of `com.amazonaws.apigatewayv2#GetRoute` from `com.amazonaws.apigatewayv2#GetRouteResult` to `com.amazonaws.apigatewayv2#GetRouteOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:5377:44
[ERROR] com.amazonaws.apigatewayv2#UpdateIntegration: Changed operation output of `com.amazonaws.apigatewayv2#UpdateIntegration` from `com.amazonaws.apigatewayv2#UpdateIntegrationResult` to `com.amazonaws.apigatewayv2#UpdateIntegrationOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:8526:53
[ERROR] com.amazonaws.apigatewayv2#UntagResource: Changed operation output of `com.amazonaws.apigatewayv2#UntagResource` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#UntagResourceResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:7709:49
[ERROR] com.amazonaws.apigatewayv2#GetIntegration: Changed operation output of `com.amazonaws.apigatewayv2#GetIntegration` from `com.amazonaws.apigatewayv2#GetIntegrationResult` to `com.amazonaws.apigatewayv2#GetIntegrationOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:4711:50
[ERROR] com.amazonaws.apigatewayv2#DeleteDomainName: Changed operation output of `com.amazonaws.apigatewayv2#DeleteDomainName` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteDomainNameResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3033:52
[ERROR] com.amazonaws.apigatewayv2#DeleteRoute: Changed operation output of `com.amazonaws.apigatewayv2#DeleteRoute` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteRouteOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3205:47
[ERROR] com.amazonaws.apigatewayv2#DeleteApiMapping: Changed operation output of `com.amazonaws.apigatewayv2#DeleteApiMapping` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteApiMappingResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2853:52
[ERROR] com.amazonaws.apigatewayv2#CreateRoute: Changed operation output of `com.amazonaws.apigatewayv2#CreateRoute` from `com.amazonaws.apigatewayv2#CreateRouteResult` to `com.amazonaws.apigatewayv2#CreateRouteOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2091:47
[ERROR] com.amazonaws.apigatewayv2#DeleteCorsConfiguration: Changed operation output of `com.amazonaws.apigatewayv2#DeleteCorsConfiguration` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteCorsConfigurationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2955:59
[ERROR] com.amazonaws.apigatewayv2#CreateIntegration: Changed operation output of `com.amazonaws.apigatewayv2#CreateIntegration` from `com.amazonaws.apigatewayv2#CreateIntegrationResult` to `com.amazonaws.apigatewayv2#CreateIntegrationOperationResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:1524:53
[ERROR] com.amazonaws.apigatewayv2#DeleteApi: Changed operation output of `com.amazonaws.apigatewayv2#DeleteApi` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteApiResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:2831:45
[ERROR] com.amazonaws.apigatewayv2#DeleteModel: Changed operation output of `com.amazonaws.apigatewayv2#DeleteModel` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteModelResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3162:47
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteResponse: Changed operation output of `com.amazonaws.apigatewayv2#DeleteRouteResponse` from `smithy.api#Unit` to `com.amazonaws.apigatewayv2#DeleteRouteResponseResponse` | ChangedOperationOutput apigatewayv2.2018-11-29.json:3299:55
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3321:62
[ERROR] com.amazonaws.apigatewayv2#GetStageRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5794:51
[ERROR] com.amazonaws.apigatewayv2#DeleteIntegrationResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3133:68
[ERROR] com.amazonaws.apigatewayv2#GetStagesRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5946:52
[ERROR] com.amazonaws.apigatewayv2#GetVpcLinksRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6178:54
[ERROR] com.amazonaws.apigatewayv2#CreateDomainNameResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1484:60
[ERROR] com.amazonaws.apigatewayv2#GetStagesResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5973:53
[ERROR] com.amazonaws.apigatewayv2#GetAuthorizersRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4354:57
[ERROR] com.amazonaws.apigatewayv2#GetVpcLinksResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6197:55
[ERROR] com.amazonaws.apigatewayv2#GetAuthorizersResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4381:58
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4736:57
[ERROR] com.amazonaws.apigatewayv2#DeleteVpcLinkResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3482:57
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationsResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5140:59
[ERROR] com.amazonaws.apigatewayv2#TagResourceRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7604:54
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationResponseResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4811:66
[ERROR] com.amazonaws.apigatewayv2#CreateIntegrationResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1722:68
[ERROR] com.amazonaws.apigatewayv2#GetDomainNamesResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4692:58
[ERROR] com.amazonaws.apigatewayv2#CreateRouteResponseResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2300:63
[ERROR] com.amazonaws.apigatewayv2#CreateApiMappingRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:813:59
[ERROR] com.amazonaws.apigatewayv2#CreateDeploymentResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1366:60
[ERROR] com.amazonaws.apigatewayv2#UpdateDomainNameResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8486:60
[ERROR] com.amazonaws.apigatewayv2#GetRouteResponsesResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5580:61
[ERROR] com.amazonaws.apigatewayv2#UpdateAuthorizerRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8140:59
[ERROR] com.amazonaws.apigatewayv2#UpdateApiMappingResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7866:60
[ERROR] com.amazonaws.apigatewayv2#GetModelTemplateResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5291:60
[ERROR] com.amazonaws.apigatewayv2#CreateDeploymentRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1336:59
[ERROR] com.amazonaws.apigatewayv2#UpdateDeploymentResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8378:60
[ERROR] com.amazonaws.apigatewayv2#DeleteModelRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3184:54
[ERROR] com.amazonaws.apigatewayv2#DeleteAuthorizerRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2934:59
[ERROR] com.amazonaws.apigatewayv2#CreateVpcLinkRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2682:56
[ERROR] com.amazonaws.apigatewayv2#UpdateModelResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9073:55
[ERROR] com.amazonaws.apigatewayv2#UpdateIntegrationRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8557:60
[ERROR] com.amazonaws.apigatewayv2#GetTagsResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6036:51
[ERROR] com.amazonaws.apigatewayv2#UpdateRouteRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9144:54
[ERROR] com.amazonaws.apigatewayv2#UpdateRouteResponseResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9336:63
[ERROR] com.amazonaws.apigatewayv2#ExportApiResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3804:53
[ERROR] com.amazonaws.apigatewayv2#DeleteStageRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3423:54
[ERROR] com.amazonaws.apigatewayv2#GetRouteRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5402:51
[ERROR] com.amazonaws.apigatewayv2#CreateAuthorizerResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1223:60
[ERROR] com.amazonaws.apigatewayv2#CreateRouteResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2247:62
[ERROR] com.amazonaws.apigatewayv2#GetDeploymentResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4446:57
[ERROR] com.amazonaws.apigatewayv2#GetRouteResponseResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5477:60
[ERROR] com.amazonaws.apigatewayv2#ImportApiResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6292:53
[ERROR] com.amazonaws.apigatewayv2#ExportApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3754:52
[ERROR] com.amazonaws.apigatewayv2#GetModelTemplateRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5270:59
[ERROR] com.amazonaws.apigatewayv2#GetRoutesRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5723:52
[ERROR] com.amazonaws.apigatewayv2#UpdateApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7899:52
[ERROR] com.amazonaws.apigatewayv2#UpdateAuthorizerResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8234:60
[ERROR] com.amazonaws.apigatewayv2#DeleteApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2899:52
[ERROR] com.amazonaws.apigatewayv2#CreateApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:885:52
[ERROR] com.amazonaws.apigatewayv2#GetRouteResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5448:59
[ERROR] com.amazonaws.apigatewayv2#GetAuthorizerRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4223:56
[ERROR] com.amazonaws.apigatewayv2#GetApiMappingResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3895:57
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3227:54
[ERROR] com.amazonaws.apigatewayv2#GetStageResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5815:52
[ERROR] com.amazonaws.apigatewayv2#DeleteCorsConfigurationRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2977:66
[ERROR] com.amazonaws.apigatewayv2#CreateIntegrationResponseResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1782:69
[ERROR] com.amazonaws.apigatewayv2#GetApisResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4179:51
[ERROR] com.amazonaws.apigatewayv2#GetDeploymentRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4425:56
[ERROR] com.amazonaws.apigatewayv2#CreateModelResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2051:55
[ERROR] com.amazonaws.apigatewayv2#UpdateStageRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9503:54
[ERROR] com.amazonaws.apigatewayv2#GetDeploymentsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4521:57
[ERROR] com.amazonaws.apigatewayv2#CreateApiResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:986:53
[ERROR] com.amazonaws.apigatewayv2#GetRouteResponsesRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5545:60
[ERROR] com.amazonaws.apigatewayv2#GetTagsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6023:50
[ERROR] com.amazonaws.apigatewayv2#GetDomainNamesRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4673:57
[ERROR] com.amazonaws.apigatewayv2#CreateStageResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2554:55
[ERROR] com.amazonaws.apigatewayv2#UpdateVpcLinkRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9714:56
[ERROR] com.amazonaws.apigatewayv2#UpdateIntegrationResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8731:68
[ERROR] com.amazonaws.apigatewayv2#TagResourceResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7627:55
[ERROR] com.amazonaws.apigatewayv2#DeleteAccessLogSettingsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2810:66
[ERROR] com.amazonaws.apigatewayv2#GetApiMappingsResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3983:58
[ERROR] com.amazonaws.apigatewayv2#ResetAuthorizersCacheRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7098:64
[ERROR] com.amazonaws.apigatewayv2#GetVpcLinkResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6085:54
[ERROR] com.amazonaws.apigatewayv2#CreateIntegrationRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1555:60
[ERROR] com.amazonaws.apigatewayv2#GetApiMappingsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3956:57
[ERROR] com.amazonaws.apigatewayv2#GetDomainNameResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4605:57
[ERROR] com.amazonaws.apigatewayv2#GetApiMappingRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3874:56
[ERROR] com.amazonaws.apigatewayv2#UpdateApiResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7992:53
[ERROR] com.amazonaws.apigatewayv2#GetDomainNameRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4592:56
[ERROR] com.amazonaws.apigatewayv2#DeleteDomainNameRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3055:59
[ERROR] com.amazonaws.apigatewayv2#UpdateDeploymentRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8347:59
[ERROR] com.amazonaws.apigatewayv2#ReimportApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6921:54
[ERROR] com.amazonaws.apigatewayv2#ImportApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6262:52
[ERROR] com.amazonaws.apigatewayv2#ReimportApiResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6959:55
[ERROR] com.amazonaws.apigatewayv2#GetAuthorizerResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4244:57
[ERROR] com.amazonaws.apigatewayv2#UpdateApiMappingRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7820:59
[ERROR] com.amazonaws.apigatewayv2#DeleteApiMappingRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2878:59
[ERROR] com.amazonaws.apigatewayv2#GetApiResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4015:50
[ERROR] com.amazonaws.apigatewayv2#UpdateRouteResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9276:62
[ERROR] com.amazonaws.apigatewayv2#CreateStageRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2467:54
[ERROR] com.amazonaws.apigatewayv2#UpdateDomainNameRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8456:59
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationResponseRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4782:65
[ERROR] com.amazonaws.apigatewayv2#GetModelResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5205:52
[ERROR] com.amazonaws.apigatewayv2#DeleteDeploymentRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3012:59
[ERROR] com.amazonaws.apigatewayv2#CreateAuthorizerRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1134:59
[ERROR] com.amazonaws.apigatewayv2#DeleteVpcLinkRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3469:56
[ERROR] com.amazonaws.apigatewayv2#GetRoutesResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5750:53
[ERROR] com.amazonaws.apigatewayv2#CreateApiMappingResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:852:60
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationResponsesRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4886:66
[ERROR] com.amazonaws.apigatewayv2#GetModelsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5331:52
[ERROR] com.amazonaws.apigatewayv2#CreateDomainNameRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:1447:59
[ERROR] com.amazonaws.apigatewayv2#CreateModelRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2005:54
[ERROR] com.amazonaws.apigatewayv2#UpdateModelRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9021:54
[ERROR] com.amazonaws.apigatewayv2#CreateRouteRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2122:54
[ERROR] com.amazonaws.apigatewayv2#GetModelRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5184:51
[ERROR] com.amazonaws.apigatewayv2#UntagResourceRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:7737:56
[ERROR] com.amazonaws.apigatewayv2#DeleteIntegrationRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3090:60
[ERROR] com.amazonaws.apigatewayv2#GetVpcLinkRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:6072:53
[ERROR] com.amazonaws.apigatewayv2#GetDeploymentsResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4548:58
[ERROR] com.amazonaws.apigatewayv2#GetModelsResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5358:53
[ERROR] com.amazonaws.apigatewayv2#UpdateVpcLinkResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9737:57
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteSettingsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3372:62
[ERROR] com.amazonaws.apigatewayv2#UpdateStageResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:9583:55
[ERROR] com.amazonaws.apigatewayv2#GetApisRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4160:50
[ERROR] com.amazonaws.apigatewayv2#UpdateIntegrationResponseResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:8798:69
[ERROR] com.amazonaws.apigatewayv2#GetApiRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4002:49
[ERROR] com.amazonaws.apigatewayv2#DeleteRouteRequestParameterRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:3270:70
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationResponsesResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:4921:67
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationsRequest: Added trait `smithy.api#input` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:5113:58
[ERROR] com.amazonaws.apigatewayv2#CreateVpcLinkResponse: Added trait `smithy.api#output` with value {} | ModifiedTrait apigatewayv2.2018-11-29.json:2720:57
[ERROR] com.amazonaws.apigatewayv2#GetIntegrationResult: Removed structure `com.amazonaws.apigatewayv2#GetIntegrationResult` | RemovedShape apigatewayv2.2018-11-29.json:4940:56
[ERROR] com.amazonaws.apigatewayv2#UpdateIntegrationResult: Removed structure `com.amazonaws.apigatewayv2#UpdateIntegrationResult` | RemovedShape apigatewayv2.2018-11-29.json:8845:59
[ERROR] com.amazonaws.apigatewayv2#CreateIntegrationResult: Removed structure `com.amazonaws.apigatewayv2#CreateIntegrationResult` | RemovedShape apigatewayv2.2018-11-29.json:1829:59
[ERROR] com.amazonaws.apigatewayv2#CreateRouteResult: Removed structure `com.amazonaws.apigatewayv2#CreateRouteResult` | RemovedShape apigatewayv2.2018-11-29.json:2340:53
[ERROR] com.amazonaws.apigatewayv2#GetRouteResult: Removed structure `com.amazonaws.apigatewayv2#GetRouteResult` | RemovedShape apigatewayv2.2018-11-29.json:5599:50
[ERROR] com.amazonaws.apigatewayv2#UpdateRouteResult: Removed structure `com.amazonaws.apigatewayv2#UpdateRouteResult` | RemovedShape apigatewayv2.2018-11-29.json:9376:53
</pre>
</details>

kstich commented on smithy-model/src/main/java/software/amazon/smithy/model/transform/ModelTransformer.java:
```diff
@@ -491,4 +492,56 @@ public Model changeShapeType(Model model, Map<ShapeId, ShapeType> shapeToType) {
+     * context.
+     *
+     * <p>If an operation's input already targets a shape marked with the {@code input}
+     * trait, then the operation and input shape is left as-is regardless of its name.
```
These conditionals still route through `renameShapeIfNeeded` in the transform ([input](https://github.com/awslabs/smithy/pull/992/files#diff-4bf748f6907e4f1457f9c288fd3e90fa9d01662d3fa84adebf3028f96aee36e5R107), [output](https://github.com/awslabs/smithy/pull/992/files#diff-4bf748f6907e4f1457f9c288fd3e90fa9d01662d3fa84adebf3028f96aee36e5R124)).
